### ISSUE TITLE
Task-Based UIs

### ISSUE THREAD

**Issue by gentledepp** [OrchardCMS/OrchardCore#882]
Hello!
I haven't been using orchard V1 for now as it was too much of a CMS then an application framework.
However, with v2 this seems to chagne. I am pretty happy about that!

What we need in our application are task-based UIs.
Like there is a contenttype task.
When the user wants to edit that task, there are different fields (or content parts) displayed depending on the state that the task is currently in.

Orchard v1 seemed to have much more a CRUD approach which is not sufficient for us.

Are there any plans to support this, or are there any best practices/approaches to solve such a task?

**Comment by Jetski5822** [comment]
Hi @gentledepp - Hmmm im not sure we have this planned. Though I feel like this could be done once workflows is finished.....

A User creates a page.... some thing happens, and that page is in a new state etc...

**Comment by agriffard** [comment]
@sfmskywalker is actively working on Workflows but I don't know if it will cover this kind of scenario.

**Comment by sfmskywalker** [comment]
Although the Workflows module itself doesn't provide anything related to task-based UIs, it would allow one to implement task-based UIs and associate them with workflows. I.e. one would have a set of Tasks, and whenever a task is updated, a workflow executes (started or resumed). Since the Workflows module supports long-running workflows, it is a great way to persist state for and provide state to each task. 

Combined with upcoming support for Forms, we could think about a new feature named Tasks that formalizes the notion of an actual task, its steps and their state. A task would contain the entire state of that task, and an associated workflow instance would drive which step to present next, including collecting any input from a given step. As part of this feature an additional set of workflow activities and JS functions would be provided to make it easy to read and update a step's state.

With all that in place, an Orchard user could:

- Define Task Definitions - Task Definitions consist of a Title, a reference to a Workflow Definition, and a reference to the current Step.
- Define Steps - "Step" is a content type that contains all sorts of input (Forms) widgets. It represents a single step in a given task.
- Define a Workflow Definition that decides what the next Step will be. This is an example of a new activity that would both perform an action as well as halting the current workflow.
- Start a new Task - Task is an "instance" of a Task Definition, like Workflow is an "instance" of a Workflow Definition. Task has a reference to its Task Definition as well as a Workflow Instance.
- Since Steps are content items with widgets, they are presented on the front-end, but it would be brilliant if we could also render a Task within the admin UI.

In addition, a workflow itself could start a new Task. This would unlock scenarios such as Content Approval Workflows. For example:

1. A user submits a Suggestion form on the front-end, which is posted to a workflow.
2. The workflow starts a new Task called "Review Suggestion" and assigns it to a Moderator user.
3. The "Review Suggestion" task kicks off its associated workflow, which:
4. Sends an email to its associated Moderator and sets the next Step.
5. The moderator clicks on a link to the task, which shows the current step.
6. The current step renders the provided suggestion and allows the moderator to Approve or Reject.
7. Clicking either Approve or Reject will trigger the associated workflow.
8. Eventually the task comes to an end.

This way, you could also imagine users logging into the admin UI and have a customized dashboard with a widget listing Tasks for the user to perform.

**Comment by sfmskywalker** [comment]
Come to think of it, a Task and a Step sounds very similar to a Workflow Instance and an Activity. Perhaps my proposal could be simplified to simply introducing a Step content type.

**Comment by sfmskywalker** [comment]
Additionally, there would be a Display Step activity that allows the user to specify which Step to display. No need for an actual Task entity to maintain state - that is the responsibility of a Workflow Instance.

Unless someone thinks there's value to being able to associate metadata to a workflow instance and formalize the notion of a User Task.

**Comment by sebastienros** [comment]
Definitely Workflows.

**Comment by sebastienros** [comment]
Closing because it's almost merged
